### ISSUE TITLE
[feat]: <rh-code-block>

### ISSUE THREAD

**Issue by eohammond** [RedHat-UX/red-hat-design-system#328]
`<rh-code-block>`

### Mockups

- [Design specs w/ options](https://xd.adobe.com/view/a64d7305-b6cb-439c-8caa-1039834fa1b8-f31e/)
- [Syntax highlighting tokens](https://xd.adobe.com/view/c4cc943e-618d-4fbe-8ab2-8e29fd135393-919f/)

### Options/Feature priorities
_Taken from Benny's comment below_

1. base style
2. color context
3. copy button (blocked on copy button)
4. resizable
5. expandable
6. line wrap (blocked on switch)
7. syntax (blocked on code tokens)

### Patterns
_To be implemented later_

1. toolbar with (switch|copy|links)
2. tabs with (switch|copy)

**Comment by zeroedin** [comment]
We would only remove `pfe-codeblock` if it did not exist in Patternfly proper.  Because there is a Patternfly version this would be a 1:1 migration.  `pfe-codeblock` would then become a 1:1 match with its counterpart in Patternfly, then the `rh-codeblock` component would extend the pfe version by applying our design system customizations.  An example of this is @bennypowers work on `pfe-modal` -> `rh-dialog` 


**pfe-modal** https://github.com/patternfly/patternfly-elements/pull/2036
and
**rh-dialog** https://github.com/RedHat-UX/red-hat-design-system/pull/288

**Comment by coreyvickery** [comment]
### Requirements

* Light and dark themes
* Accessible color styles within both themes
* Copy to clipboard button
* Expandable `Show more` button or function

### Current design to work from

https://xd.adobe.com/view/fb466c74-091f-44dd-9a15-c66f80783e5b-ae15/

### Relevant chats

* `We also have copy to clipboard as a separate component` (MC)
* `I think we could implement a cleaner version of the expandable section` (MC)
* `It feels like the "show more" should be in a button maybe, so it doesn't feel awkward or lost next to the code` (MC)
* `Actually, "copy" is part of the PFE implementation too—on hover` (MC)

**Comment by methomps** [comment]
Cloud.redhat.com's use of codeblock is not reliant on our implementation. AKA we are not a blocker.

**Comment by coreyvickery** [comment]
@methomps Possible to provide any screenshots from that environment? I'm having trouble logging into my account.

**Comment by coreyvickery** [comment]
@eohammond If you're able to provide, yes.

**Comment by coreyvickery** [comment]
@methomps This is being reviewed by leads from other teams.

https://xd.adobe.com/view/7519da71-4ac1-440e-a7f2-bf25de199cd1-204a/

**Comment by coreyvickery** [comment]
@methomps We are still blocked on this until we sync with the other design system leads to see what research and/or design they have already done, will try to move this forward early next week.

**Comment by methomps** [comment]
Probably a good candidate to test our new design process once we've defined that process a bit more with the leads collaborative.

**Comment by bennypowers** [comment]
1:1: https://github.com/patternfly/patternfly-elements/issues/2093

Also relevant: https://github.com/RedHat-UX/red-hat-design-tokens/issues/16

**Comment by bennypowers** [comment]
Ideally we wouldn't have to create this at all, and rather use a [prism or highlight.js syntax theme](https://github.com/RedHat-UX/red-hat-design-tokens/issues/16). Talking it over with @mwcz, we agreed that doing code highlighting client-side is not ideal. This could be useful for certain cases like cms or places where it's cumbersome to server-render code, but shouldn't be the first choice. 

On the implementation side, if we do implement this (and pfe-codeblock), I'd prefer to either fork or white-label the existing `<zero-md>` element.

cc @markcaron @heyMP

**Comment by markcaron** [comment]
@marionnegp and @bennypowers, I believe we need to touch base on this.

**Comment by coreyvickery** [comment]
Current design:

https://xd.adobe.com/view/4c659461-a169-4e21-80bd-e4039aff3e7b-1936/

**Comment by coreyvickery** [comment]
Syntax highlighting:

https://xd.adobe.com/view/c4cc943e-618d-4fbe-8ab2-8e29fd135393-919f/

**Comment by bennypowers** [comment]
priorities:
1. base style
1. color context
1. copy button
1. resizable
1. expandable
1. syntax

dependencies
1. switch
2. highlighting
3. toolbar/panel

**Comment by markcaron** [comment]
For line highlighting, see [`<mark>` element in MDN Web Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/mark#accessibility_concerns)

**Comment by bennypowers** [comment]
@coreyvickery grey background and border colours `#F5F5F5` and `#D2D2D2` aren't token values, please advise

**Comment by coreyvickery** [comment]
@bennypowers I forgot the token names because I’m mobile, but F5 -> F2 and D2 -> C7. Those hex values should map to tokens.

**Comment by bennypowers** [comment]
Thanks I'll check tomorrow and get back to you if I have more questions 🙏

**Comment by bennypowers** [comment]
can you elaborate on the resize behaviour?
What happens before, during, and after the resize action (i.e. drag handle)
it looks in the mockup like prior to resizing the single long line of text is broken into two, but arbitrarily, while there's also a horizontal scrollbar, whereas after resize, the text wraps at the element width. is that correct? is there a prescribed width for the scrollable container before resize?

**Comment by bennypowers** [comment]
yes resize can be in either both or none directions: https://developer.mozilla.org/en-US/docs/Web/CSS/resize

I'd like to see a rough ux demo of the intention here. it's not clear to me yet how much of this we can accomplish easily with css

**Comment by bennypowers** [comment]
perhaps we can workshop this for a bit at office hours

**Comment by bennypowers** [comment]
could you export svgs for the action icons please?

**Comment by bennypowers** [comment]
I did a little work on copy buttons for code block (only) this morning, based on the xd here, but i'm going to revert that work pending discussion in https://github.com/RedHat-UX/red-hat-design-system/issues/423

tl;dr: we can't use tooltip or button text for copy state, should use alert, plus perhaps role=status cc @nikkimk for comments and corrections, with thanks.

**Comment by bennypowers** [comment]
## Buttons
The button styles in the XD are mostly bespoke (particularly the "action" buttons - the ones superimposed in the top right of the code block. do these represent new button variants, or should I continue to just hack them into bespoke variants private to code block?

## Panel/Card/Toolbar
The upper toolbar, is this a component? Are there or will there be other components that fit into this "toolbar with content" deal?

**Comment by bennypowers** [comment]
## Callouts
can they go in a 'gutter' on the left (inline start) instead of at the end? this would be easier to implement and read (imo) for word-wrapped snippets

**Comment by bennypowers** [comment]
Today at Office hours we took the decision to release a very basic version of code-block with bellsprout.

features we'll push off to bulbasaur:
- new variants of `<rh-button>`: icon button / FAB for copy
- syntax highlighting
- callout badges
- toolbar/container

We also have some usability questions we'd like to address:
- should copy button always be a FAB (i.e. floating over the codeblock, not slotted into the toolbar)
- is combining tabs with links/actions a usability problem for sighted keyboard users?

SO, basically the bellsprout version will be a gray box that formats a script of text
